### ISSUE TITLE
Peer Review Feedback

### ISSUE THREAD

**Issue by andrewdes** [OpenLiberty/draft-guide-istio#14]
- [ ] Working instruction and sample code
- [x] Quality of code (best coding practices)
- [x] Quality and presentation of guide according to Structure and Style Guideline
- [x] Consistency of guide with template and other guides
- [ ] Examples with the right outcomes are provided
- [x] Example with the wrong outcomes, if any, are provided
- [x] Check that all licensing statements are properly stated in all files
- [x] Check that the pom.xml, server.xml, etc files are clean up
- [x] Check that the directories are properly structured
- [x] Test the guide on 3 platforms: Mac, Win, Linux
- [x] If any URL visits, try them on different browsers: Firefox, Chrome, Safari
- [x] If any URL visits, try curl command where applicable
- [ ] Ensure that the automation tests are able to run with continuous daily build

----

### General
- [x] Title displayed on guide is wrong
- [x] Blue-green deployments need to be further explained
- [x] Provide an overall picture of what the user will be accomplishing in this guide
- [ ] Do the .yaml files and set-live scripts need licensing statements?
- [x] There is a large chunk of commented out code in the pom.xml file. Is this necessary?
- [ ] Test automation script needs to be updated (daily build)

### What you'll learn
- [x] I feel like the opening 3 paragraphs could be expanded to include more information
- [x] `kubectl` CLI -> Should the first occurrence of CLI be expanded (command-line interface)?
- [x] "It works by deploying istio control plane" Should Istio be capitalized?
- [x] "It’s specialized for a service mesh architecture since it offers many useful features." Is the service mesh architecture something that should be further expanded upon? Or are we assuming users will be familiar with it already?
- [x] "Features it offers such as load balancing and support for distributed tracing **help creating a service mesh**." -> "help create a service mesh" sounds better
- [x] "While Istio supports Kubernetes and that will be the focus of this guide." -> This sentence sounds incomplete.
- [x] "While Istio supports Kubernetes and that will be the focus of this guide. Note that Istio can also be used with other environments such as Docker Compose" -> Perhaps you can combine these: "While the focus of this guide will be Istio with Kubernetes, note that Istio can also be used with other environments such as Docker Compose."
- [x] "Istio has many features. Some of these features are traffic shifting, request routing, access control, and distributed tracing. The focus of this guide will be traffic shifting."  -> Try combining/re-wording these to flow better. Maybe something like: "Although Istio has many features such as traffic shifting, request routing, access control and distributed tracing, the focus of this guide will be traffic shifting." 
- [x] In the **Why Istio** section, you mention that Istio provides a suite of features to manage Routing, Telemetry, and Security. However, in the following paragraph you only mention use cases for Telemetry and Security.
- [x] "You will use Istio’s traffic shifting features to implement blue-green deployments." -> This sentence seems out of place and should be expanded or moved so that it flows better.

### Prerequisites
- [ ] Links should be the title of the page they user will be brought to, not full links (i.e. https://docs.docker.com/docker-for-windows/#kubernetes.)
- [ ] "Use the Docker Desktop" -> "Use Docker Desktop"

### Starting and preparing your cluster for deployment
- [ ] "You do not need to do any other step." -> "You do not need to do any other steps" OR "You do not need to do another step"

### Deploying Istio
- [x] Not all Istio deployments became available. However, after increasing memory in Docker preferences from 2gb -> 4gb all deployments became available. This might be worth mentioning somewhere.

### Deploying version 1 of the service to the cluster
- [x] You tell the user to run `kubectl apply -f traffic.yaml` then immediately display the output. Perhaps separate the output and command with a sentence or two (maybe a brief explanation of what the command they ran does, "you can see that your resources are created: ", etc.).
- [x] "Once all the deployments are available,..." -> This paragraph might work better after running `kubectl get deployments` and verifying your deployments are available.
- [x] "If curl is unavailable then use Postman." -> This might flow better if moved to after the curl command is displayed.

### Deploying version 2 of the service to the cluster
- [x] "Test the updated microservice by making requests to the testing site" -> What testing site? I feel like the overall picture of what is happening/what the user is accomplishing in this section needs to be clearer. 
- [x] "modify the traffic.yaml file to shift 100 percent of the live traffic to the green deployment." -> Perhaps you can explain what is being modified/what needs to be done to the file in order to shift the live traffic to the green deployment.
- [x] This section overall was just a lot of "execute this command" without much explanation. I found it difficult to connect this section with the overall purpose of the guide. I think if blue-green deployments are further explained in the intro, this section might make more sense. In addition, perhaps explaining that there is a test site and a live site might be useful, maybe in the intro when talking about the hello microservice. 

### Automating the blue-green switch
- [x] Once again this section was essentially just "execute this command" with little to no explanation. I understand that most of these commands are simply repeated. However, for the set-live script, there wasn't much explanation other than the fact that it is a script that automates the traffic shifting. 

### Testing microservices that are running on Kubernetes
- [x] "Create the endpoint test in the `src/test/java/it/io/openliberty/guides/rest/EndpointTest.java` file:" -> Doesn't appear to be consistent with other guides
- [ ] The test could be explained with more detail
- [x] "The cluster.ip and port parameters refer to the IP address and port for the Istio gateway." If the user is on Windows/Mac, they will not use these parameters
- [x] "If the tests pass, then you should see output similar to **this**:" -> "If the tests pass, then you should see output similar to **the following**:"
- [x] "Roll back to the old version of the microservice and run the tests again." -> "Next, roll back to the old version of the microservice and run the tests again."
- [x] "Observe that the tests fail because the version number that the tests expect and the version of the running microservice are now different." -> This might work better before the tests so the user expects to see failed test results. Maybe something like: "Next, roll back to the old version of the microservice and run the tests again. Notice how when you run the tests this time they fail..."

**Comment by andrewdes** [comment]
When running through the guide on Windows I ran into the following problems: 

---

When running `mvn clean package` in the **Deploying version 2 of the hello microservice**, I get the following error: 

```
[ERROR] Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on <ip_address>:53: read udp <ip_address_2>:38405-><ip_address>:53: i/o timeout
[WARNING] An attempt failed, will retry 1 more times
org.apache.maven.plugin.MojoExecutionException: Could not build image
    at com.spotify.plugin.dockerfile.BuildMojo.buildImage (BuildMojo.java:185)
    at com.spotify.plugin.dockerfile.BuildMojo.execute (BuildMojo.java:105)
    at com.spotify.plugin.dockerfile.AbstractDockerMojo.tryExecute (AbstractDockerMojo.java:259)
    at com.spotify.plugin.dockerfile.AbstractDockerMojo.execute (AbstractDockerMojo.java:248)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:90)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:55)
    at java.lang.reflect.Method.invoke (Method.java:508)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
Caused by: com.spotify.docker.client.exceptions.DockerException: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on <ip_address>:53: read udp <ip_address_2>:38405-><ip_address>:53: i/o timeout
    at com.spotify.plugin.dockerfile.LoggingProgressHandler.handleError (LoggingProgressHandler.java:105)
    at com.spotify.plugin.dockerfile.LoggingProgressHandler.progress (LoggingProgressHandler.java:63)
    at com.spotify.docker.client.DefaultDockerClient$BuildProgressHandler.progress (DefaultDockerClient.java:305)
    at com.spotify.docker.client.ProgressStream.tail (ProgressStream.java:77)
    at com.spotify.docker.client.DefaultDockerClient$ResponseTailReader.call (DefaultDockerClient.java:2734)
    at com.spotify.docker.client.DefaultDockerClient$ResponseTailReader.call (DefaultDockerClient.java:2718)
    at java.util.concurrent.FutureTask.run (FutureTask.java:277)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1160)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:635)
    at java.lang.Thread.run (Thread.java:811)
[INFO] Building Docker context C:\Users\Desktop\draft-guide-istio\start
[INFO]
[INFO] Image will be built as hello:2.0-SNAPSHOT
[INFO]
[INFO] Step 1/6 : FROM open-liberty
[INFO]
[INFO] Pulling from library/open-liberty
[INFO] Digest: sha256:49c1545ef3a748a72ae0e5ca47bfdfa6051e2704addb71dc46471b6a11a6f818
[INFO] Status: Image is up to date for open-liberty:latest
[INFO]  ---> af81d166e48d
[INFO] Step 2/6 : COPY target/liberty/wlp/usr/servers/defaultServer/bootstrap.properties /config
[INFO]
[INFO]  ---> 35bc1d33d694
[INFO] Step 3/6 : COPY target/liberty/wlp/usr/servers/defaultServer/server.env /config
[INFO]
[INFO]  ---> 5043355e67b8
[INFO] Step 4/6 : COPY src/main/liberty/config/server.xml /config
[INFO]
[INFO]  ---> 3009b17f3505
[INFO] Step 5/6 : COPY target/*.war /config/apps/hello.war
[INFO]
[INFO]  ---> 4db5ac226ffc
[INFO] Step 6/6 : EXPOSE 9080
[INFO]
[INFO]  ---> Running in 41be47e9c569
[INFO] Removing intermediate container 41be47e9c569
[INFO]  ---> 33c91ffb34ee
[INFO] Successfully built 33c91ffb34ee
[INFO] Successfully tagged hello:2.0-SNAPSHOT
[INFO]
[INFO] Detected build of image with id 33c91ffb34ee
[INFO] Building jar: C:\Users\Desktop\draft-guide-istio\start\target\hello-2.0-SNAPSHOT-docker-info.jar
[INFO] Successfully built hello:2.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:13 min
[INFO] Finished at: 2019-01-04T10:00:29-05:00
[INFO] ------------------------------------------------------------------------
```

Note that the build was still successful and I was able to curl correctly afterwards. 

---

In the **Automating the blue-green switch** the `set-live.bat` script produces an error when running `set-live.bat blue`:
```
Setting blue as live...
error: error converting YAML to JSON: yaml: line 8: did not find expected key
```

**Comment by proubatsis** [comment]
> When running mvn clean package in the Deploying version 2 of the hello microservice, I get the following error:

I cannot reproduce this issue.
> In the Automating the blue-green switch the set-live.bat script produces an error when running set-live.bat blue:

Fixed in #18

**Comment by proubatsis** [comment]
@andrewdes If all looks good then please close this issue.
